File: clients/go/kimmydb/watch.go
package kimmydb

import (
	"context"
	"encoding/json"
	"fmt"
	"iter"
	"net/http"
	"strings"
	"time"

	"github.com/coder/websocket"
)

// ChangeEvent is one event from a collection.
type ChangeEvent struct {
	// Operation is insert, update, replace, delete, uniqueViolation or
	// invalidate.
	Operation string
	// ResumeToken is where to resume. Empty on invalidate, which cannot be
	// resumed past.
	ResumeToken string
	// Raw is the whole event, for everything the fields above do not name.
	Raw map[string]any
}

// DocumentID returns the changed document's _id, when the event has one.
func (e ChangeEvent) DocumentID() any {
	key, _ := e.Raw["documentKey"].(map[string]any)
	return key["_id"]
}

// FullDocument returns the post-image, when it was asked for and the event
// carried one.
//
// An oversized event drops it and still arrives, so absence does not mean the
// document is gone.
func (e ChangeEvent) FullDocument() map[string]any {
	document, _ := e.Raw["fullDocument"].(map[string]any)
	return document
}

// IsInvalidate reports whether the stream cannot continue past this event.
func (e ChangeEvent) IsInvalidate() bool { return e.Operation == "invalidate" }

// InvalidateReason returns why the stream ended: CollectionDropped,
// ConsumerLagged or ResumeTokenExpired. Treat an unrecognized value as "this
// stream is over", which is what every value means.
func (e ChangeEvent) InvalidateReason() string {
	reason, _ := e.Raw["reason"].(string)
	return reason
}

// WatchOptions configures a change stream.
type WatchOptions struct {
	// ResumeAfter continues immediately after a token from a previous stream.
	ResumeAfter string
	// FromStart replays from the beginning of the retained oplog — for this
	// collection's current incarnation. A collection dropped and recreated
	// under the same name reuses its id, and its predecessor's entries are not
	// its history.
	FromStart bool
	// FullDocument includes the whole document on every event that has one.
	FullDocument bool
}

// Watch follows a collection's changes.
//
//	for event, err := range db.Watch(ctx, "shop", "orders", kimmydb.WatchOptions{}) {
//	    if err != nil { return err }
//	    fmt.Println(event.Operation, event.DocumentID())
//	}
//
// The iterator reconnects on its own, resuming from the last token it saw, and
// ends for exactly two reasons: an invalidate — the stream is over, and the
// event says why — or a failure it could not recover from, which arrives as
// the error in the loop.
//
// Cancel the context to stop early.
func (c *Client) Watch(ctx context.Context, db, collection string, options WatchOptions) iter.Seq2[ChangeEvent, error] {
	return func(yield func(ChangeEvent, error) bool) {
		stream := &changeStream{
			client:     c,
			db:         db,
			collection: collection,
			options:    options,
			resume:     options.ResumeAfter,
		}
		defer stream.close()

		for {
			if stream.socket == nil {
				if err := stream.reconnect(ctx); err != nil {
					yield(ChangeEvent{}, err)
					return
				}
			}

			_, raw, err := stream.socket.Read(ctx)
			if err != nil {
				if ctx.Err() != nil {
					return
				}
				// A dropped socket, a close frame, a timeout — all three mean
				// the same thing: reconnect and resume.
				stream.close()
				continue
			}

			var body map[string]any
			if err := json.Unmarshal(raw, &body); err != nil {
				yield(ChangeEvent{}, fmt.Errorf("kimmydb: an event was not JSON: %w", err))
				return
			}

			operation, _ := body["operationType"].(string)
			token, _ := body["resumeToken"].(string)
			event := ChangeEvent{Operation: operation, ResumeToken: token, Raw: body}
			// Recorded before the event is handed over, so a caller that stops
			// iterating mid-event resumes at the last one it actually saw.
			if token != "" {
				stream.resume = token
			}

			if !yield(event, nil) {
				return
			}
			if event.IsInvalidate() {
				return
			}
		}
	}
}

type changeStream struct {
	client     *Client
	db         string
	collection string
	options    WatchOptions
	resume     string
	socket     *websocket.Conn
}

func (s *changeStream) close() {
	if s.socket != nil {
		_ = s.socket.Close(websocket.StatusNormalClosure, "")
		s.socket = nil
	}
}

// reconnect dials, backing off, resuming from the last token seen.
func (s *changeStream) reconnect(ctx context.Context) error {
	const attempts = 5
	delay := 100 * time.Millisecond

	var last error
	for attempt := range attempts {
		if attempt > 0 {
			select {
			case <-ctx.Done():
				return ctx.Err()
			case <-time.After(delay):
			}
			if delay *= 2; delay > 5*time.Second {
				delay = 5 * time.Second
			}
		}
		err := s.dial(ctx)
		if err == nil {
			return nil
		}
		// A resume point that cannot be continued from cannot be waited out:
		// retrying the same token loops forever, and the caller has to decide
		// what to do about the gap.
		if apiErr, ok := err.(*APIError); ok && apiErr.Code == "resume_token_expired" {
			return err
		}
		last = err
	}
	return last
}

func (s *changeStream) dial(ctx context.Context) error {
	if err := s.client.authenticate(ctx); err != nil {
		return err
	}
	endpoint := s.client.primary()
	target := strings.Replace(strings.Replace(endpoint, "https://", "wss://", 1), "http://", "ws://", 1)
	url := target + fmt.Sprintf("/v1/db/%s/coll/%s/watch%s", s.db, s.collection, s.query())

	header := http.Header{}
	if token := s.client.Token(); token != "" {
		header.Set("Authorization", "Bearer "+token)
	}

	// Through the client's own *http.Client: the change stream inherits the
	// same TLS configuration, proxy and transport as every other request
	// rather than dialling with a second one that can drift.
	socket, response, err := websocket.Dial(ctx, url, &websocket.DialOptions{
		HTTPClient: s.client.http,
		HTTPHeader: header,
	})
	if err != nil {
		// The server refuses before upgrading with the ordinary error
		// envelope, so the reason survives if it can be recovered.
		if response != nil {
			defer response.Body.Close()
			raw := make([]byte, 4096)
			n, _ := response.Body.Read(raw)
			return errorFrom(response.StatusCode, raw[:n], retryAfterOf(response))
		}
		return &TransportError{Endpoint: endpoint, Err: err}
	}
	// A change stream is idle by design — sometimes for hours — so the read
	// limit matters more than a timeout would.
	socket.SetReadLimit(8 << 20)
	s.socket = socket
	return nil
}

func (s *changeStream) query() string {
	parts := make([]string, 0, 2)
	// A resume point learned while streaming wins over the configured one: it
	// is where this stream actually got to. Resuming from the configured point
	// would replay everything since, which for a stream that has been up for a
	// day is a day of events delivered twice.
	if s.resume != "" {
		parts = append(parts, "resume_after="+s.resume)
	} else if s.options.FromStart {
		parts = append(parts, "from_start=true")
	}
	if s.options.FullDocument {
		parts = append(parts, "full_document=true")
	}
	if len(parts) == 0 {
		return ""
	}
	return "?" + strings.Join(parts, "&")
}

File: clients/python/src/kimmydb/watch.py
"""Change streams, and surviving the socket dropping.

A change stream is a WebSocket that stays open for as long as the application
wants events — which is longer than networks stay up. So the interesting part
of this module is not opening one; it is what happens after it closes.

**Resume tokens are portable across nodes**, verified on a real cluster, so a
reconnect may land somewhere else and continue correctly. That is what makes
automatic reconnection safe here where it would not be in a system whose
cursors belong to a session on one machine.
"""

from __future__ import annotations

import json as jsonlib
import time
from typing import TYPE_CHECKING, Any, Dict, Iterator, List, Optional
from urllib.parse import urlencode

from websockets.sync.client import connect as ws_connect

from .errors import KimmyError

if TYPE_CHECKING:  # pragma: no cover
    from .client import Client


class ChangeEvent:
    """One event from a collection."""

    def __init__(self, raw: Dict[str, Any]) -> None:
        self.raw = raw
        #: ``insert``, ``update``, ``replace``, ``delete``, ``uniqueViolation``
        #: or ``invalidate``.
        self.operation: str = raw.get("operationType", "")
        #: Where to resume. Absent on ``invalidate``, which cannot be resumed
        #: past.
        self.resume_token: Optional[str] = raw.get("resumeToken")

    @property
    def document_id(self) -> Any:
        """The changed document's ``_id``, when the event has one."""
        return (self.raw.get("documentKey") or {}).get("_id")

    @property
    def full_document(self) -> Optional[Dict[str, Any]]:
        """The post-image, when it was asked for and the event carried one.

        **An oversized event drops it and still arrives**, so absence does not
        mean the document is gone.
        """
        return self.raw.get("fullDocument")

    @property
    def is_invalidate(self) -> bool:
        """Whether the stream cannot continue past this event."""
        return self.operation == "invalidate"

    def __repr__(self) -> str:  # pragma: no cover - debugging aid
        return f"<ChangeEvent {self.operation} _id={self.document_id!r}>"


class ChangeStream:
    """A live change stream that reconnects on its own.

    ::

        for event in db.watch("shop", "orders", full_document=True):
            print(event.operation, event.document_id)

    Iteration ends for exactly two reasons: an ``invalidate``, which is the
    collection going away, and a resume point that has fallen past the
    retention horizon — which cannot be waited out, because retrying the same
    token loops forever.
    """

    #: How many times to reconnect before giving up.
    ATTEMPTS = 5

    def __init__(
        self,
        client: "Client",
        db: str,
        collection: str,
        *,
        resume_after: Optional[str] = None,
        from_start: bool = False,
        full_document: bool = False,
    ) -> None:
        self._client = client
        self._db = db
        self._collection = collection
        self._configured_resume = resume_after
        self._from_start = from_start
        self._full_document = full_document
        self._resume: Optional[str] = None
        self._socket: Any = None
        self._ended = False
        # Connected here rather than on first iteration, and it matters more
        # than it looks. Python's natural shape is a lazy generator, which
        # would open the socket when the caller first reads — so anything
        # written between `watch()` and that first read would be missed, with
        # no error and nothing to see. Found by a test that wrote a document
        # immediately after opening a stream and then waited forever for it.
        self._reconnect()

    @property
    def resume_token(self) -> Optional[str]:
        """The token this stream would resume from.

        Worth storing if the application will restart: it is portable, so the
        next run may hand it to a different node.
        """
        return self._resume or self._configured_resume

    def close(self) -> None:
        if self._socket is not None:
            try:
                self._socket.close()
            finally:
                self._socket = None

    def __enter__(self) -> "ChangeStream":
        return self

    def __exit__(self, *_: object) -> None:
        self.close()

    def __iter__(self) -> Iterator[ChangeEvent]:
        while not self._ended:
            if self._socket is None:
                self._reconnect()
            try:
                message = self._socket.recv()
            except Exception:
                # A dropped socket, a timeout, a close frame — all three mean
                # the same thing: reconnect and resume.
                self.close()
                self._reconnect()
                continue

            if isinstance(message, bytes):  # pragma: no cover - server sends text
                message = message.decode()
            event = ChangeEvent(jsonlib.loads(message))
            # Recorded before the event is handed over, so a caller that stops
            # iterating mid-event resumes at the last one it actually saw.
            if event.resume_token:
                self._resume = event.resume_token
            if event.is_invalidate:
                self._ended = True
            yield event

    # -- connecting --------------------------------------------------------

    def _query(self) -> str:
        params: List[tuple] = []
        # A resume point learned while streaming wins over the configured one:
        # it is where this stream actually got to. Resuming from the configured
        # point would replay everything since, which for a stream that has been
        # up for a day is a day of events delivered twice.
        resume = self._resume or self._configured_resume
        if resume:
            params.append(("resume_after", resume))
        elif self._from_start:
            params.append(("from_start", "true"))
        if self._full_document:
            params.append(("full_document", "true"))
        return f"?{urlencode(params)}" if params else ""

    def _reconnect(self) -> None:
        delay = 0.1
        for attempt in range(self.ATTEMPTS):
            if attempt:
                time.sleep(delay)
                delay = min(delay * 2, 5.0)
            try:
                self._connect()
                return
            except KimmyError as e:
                # A resume point past the retention horizon cannot be waited
                # out, and the caller has to decide what to do about the gap.
                if e.code == "resume_token_expired":
                    self._ended = True
                    raise
                if attempt + 1 == self.ATTEMPTS:
                    raise
            except Exception:
                if attempt + 1 == self.ATTEMPTS:
                    raise

    def _connect(self) -> None:
        endpoint = self._client.endpoints[0]
        token = self._client.token
        url = (
            endpoint.replace("https://", "wss://", 1).replace("http://", "ws://", 1)
            + f"/v1/db/{self._db}/coll/{self._collection}/watch"
            + self._query()
        )
        headers = {"Authorization": f"Bearer {token}"} if token else {}
        try:
            self._socket = ws_connect(url, additional_headers=headers)
        except Exception as e:
            # The server refuses before upgrading with the ordinary error
            # envelope, so the reason survives if it can be recovered.
            status = getattr(getattr(e, "response", None), "status_code", None)
            if status is not None:
                body = getattr(e.response, "body", None)
                parsed = None
                if body:
                    try:
                        parsed = jsonlib.loads(body)
                    except ValueError:
                        parsed = None
                raise KimmyError.from_response(status, parsed) from e
            raise
